### PR TITLE
Restoring state after inference only for stateful models

### PR DESCRIPTION
**Pull request by alexsu52**
### Changes

Restoring state after inference only for stateful models

### Reason for changes

Inference optimization

### Related tickets

131141

### Tests

test_examples 230

### CODE REVIEW

File: nncf/quantization/algorithms/accuracy_control/openvino_backend.py [hunk 1]
```diff
@@ -101,5 +124,7 @@ def get_model_size(model: ov.Model) -> int:
     # Preparation of model
 
     @staticmethod
-    def prepare_for_inference(model: ov.Model) -> ov.CompiledModel:
-        return ov.compile_model(model)
+    def prepare_for_inference(model: ov.Model) -> OVPreparedModel:
```
Comment (andrey-churkin) [inline nncf/quantization/algorithms/accuracy_control/openvino_backend.py:127]: This method was removed from `AccuracyControlAlgoBackend` class

File: nncf/quantization/algorithms/accuracy_control/openvino_backend.py [hunk 2]
```diff
@@ -26,10 +27,32 @@
...
+
+class OVPreparedModel(PreparedModel):
+    """
+    Implementation of the `PreparedModel` for OpenVINO backend.
+    """
+
+    def __init__(self, model: ov.Model):
+        self.stateful = model_has_state(model)
+        self.compiled_model = ov.compile_model(model)
+        self.engine = None
```
Comment (andrey-churkin) [inline nncf/quantization/algorithms/accuracy_control/openvino_backend.py:46]:
```suggestion
        self._stateful = model_has_state(model)
        self._compiled_model = ov.compile_model(model)
        self._engine = None
```

File: nncf/quantization/algorithms/accuracy_control/openvino_backend.py [hunk 3]
```diff
@@ -26,10 +27,32 @@
...
+
+    @property
+    def model_for_inference(self) -> ov.CompiledModel:
+        return self.compiled_model
+
+    def get_engine(self) -> OVCompiledModelEngine:
```
Comment (andrey-churkin) [inline nncf/quantization/algorithms/accuracy_control/openvino_backend.py:52]: Maybe it also makes sense to decorate this method as a property

**Review by andrey-churkin — CHANGES_REQUESTED** [review]

File: nncf/quantization/algorithms/accuracy_control/openvino_backend.py (hunk 1 above)
Comment (alexsu52) [reply nncf/quantization/algorithms/accuracy_control/openvino_backend.py]: removed

File: nncf/quantization/algorithms/accuracy_control/openvino_backend.py (hunk 2 above)
Comment (alexsu52) [reply nncf/quantization/algorithms/accuracy_control/openvino_backend.py]: done

File: nncf/quantization/algorithms/accuracy_control/openvino_backend.py (hunk 3 above)
Comment (alexsu52) [reply nncf/quantization/algorithms/accuracy_control/openvino_backend.py]: Make sense

File: nncf/quantization/algorithms/accuracy_control/evaluator.py [hunk 4]
```diff
@@ -279,11 +276,12 @@ def collect_values_for_each_item_using_model_for_inference(
...
             # Collect metrics for each item
             values_for_each_item = [
-                self._validation_fn(model_for_inference, [data_item])[0] for data_item in dataset.get_data(indices)
+                self._validation_fn(prepared_model.model_for_inference, [data_item])[0]
+                for data_item in dataset.get_data(indices)
             ]
         else:
             # Collect outputs for each item
-            engine = EngineFactory.create(model_for_inference)
+            engine = prepared_model.get_engine()
```
Comment (andrey-churkin) [inline nncf/quantization/algorithms/accuracy_control/evaluator.py:284]:
```suggestion
            engine = prepared_model.engine
```

**Comment by andrey-churkin** [comment]
Could you please confirm that yolov8 sample (OpenVINO backend) works with your changes?

**Review by andrey-churkin — CHANGES_REQUESTED** [review]

**Comment by alexsu52** [comment]
> Could you please confirm that yolov8 sample (OpenVINO backend) works with your changes?

I have run test_examples, build 230.

File: nncf/quantization/algorithms/accuracy_control/evaluator.py (hunk 4 above)
Comment (alexsu52) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py]: done

**Review by andrey-churkin — APPROVED** [review]

**Pull request merged**